### PR TITLE
AP_HAL: AnalogSource: return bool for set pin

### PR DESCRIPTION
**Pull request by IamPete1**
Allows to check if a valid ADC pin has been selected.

### CODE REVIEW

File: libraries/AP_HAL_ChibiOS/AnalogIn.cpp [hunk 1]
```diff
@@ -140,11 +140,22 @@ float AnalogSource::voltage_latest()
...
 
-void AnalogSource::set_pin(uint8_t pin)
+bool AnalogSource::set_pin(uint8_t pin)
 {
     if (_pin == pin) {
-        return;
+        return true;
     }
+    bool found_pin = false;
+    for (uint8_t i=0; i<ADC_GRP1_NUM_CHANNELS; i++) {
```
Comment (IamPete1) [inline libraries/AP_HAL_ChibiOS/AnalogIn.cpp:149]: I also need to check for the `ANALOG_SERVO_VRSSI_PIN` special case.

File: libraries/AP_WindVane/AP_WindVane_Analog.cpp [hunk 2]
```diff
@@ -26,7 +26,10 @@ AP_WindVane_Analog::AP_WindVane_Analog(AP_WindVane &frontend) :
 
 void AP_WindVane_Analog::update_direction()
 {
-    _dir_analog_source->set_pin(_frontend._dir_analog_pin);
+    if (!_dir_analog_source->set_pin(_frontend._dir_analog_pin)) {
+        // pin invalid, don't have health montering to report yet
```
Comment (hendjoshsr71) [inline libraries/AP_WindVane/AP_WindVane_Analog.cpp:30]: few more below :)

```suggestion
        // pin invalid, don't have health monitoring to report yet
```

Comment (IamPete1) [reply libraries/AP_WindVane/AP_WindVane_Analog.cpp]: fixed, thanks

File: libraries/AP_HAL_ChibiOS/AnalogIn.cpp (hunk 1 above)
Comment (IamPete1) [reply libraries/AP_HAL_ChibiOS/AnalogIn.cpp]: fixed

**Comment by TunaLobster** [comment]
Would be nice to update the example scripts too.

**Comment by IamPete1** [comment]
> Would be nice to update the example scripts too.

done

**Comment by tridge** [comment]
needs testing on real stm32 hw, thanks!

**Comment by IamPete1** [comment]
Tested with example script and battery monitor on CubeOrange.

**Comment by peterbarker** [comment]
There's a flaw with the plan here.

I've got a case where the battery monitor has been set to analogue voltage+current, but the voltage pin is -1.

We don't call set_pin in the constructor, and even if we did there's no good way to return failure in there.

Eventual effects are that we seemingly get a random voltage measurement (CubeBlack).  I haven't been able to work out why that might be as get-reading code in AnalogIn.cpp is looking through valid pin configurations to work out which to read....

Yes, this is a misconfiguration.  Returning random data where we don't know where it's coming from is bad :-)

**Comment by IamPete1** [comment]
@peterbarker it should report unhealthy in that case.

**Pull request merged**